core/System.carp

```diff
@@ -56,6 +56,9 @@
   (register EROFS Int "EROFS")
   (register EINTR  Int "EINTR")
 
-  (doc exit "exits the program with the supplied exit code (thin wrapper for the C standard library).")
-  (register exit (Fn [Int] ()) "exit")
+  (doc      exit! "exits the program with the supplied exit code (thin wrapper for the C standard library).")
+  (register exit! (Fn [Int] ()) "exit")
```

Comment:
I would actually just keep the `exit` one, while the signature is slightly inaccurate, it means you can exit whenever, and it doesn't matter what it returns because it's never used.

Reply:
Done.

Frankly I don't like the signature of (System.exit) but there's a limit on how long I want to spent on this topic.

Reply:
Fair enough.

Reply:
Sorry, it's not resolved yet. I stumbled upon the original reason for changing the interface:

With the original (System.exit) I get linker errors on Windows (about missing WinMain):
1) llvm:
```
LINK : fatal error LNK1561: Einstiegspunkt muss definiert werden.
clang: error: linker command failed with exit code 1561 (use -v to see invocation)
```

2) gcc:
```
D:/.../lib/libmingw32.a(lib64_libmingw32_a-crt0_c.o):crt0_c.c:(.text.startup+0x2e): undefined reference to `WinMain'
collect2.exe: error: ld returned 1 exit status
```

Using ```(ignore (System.exit...))``` does not improve things, now Carp compilation fails with:
```

I encountered an error when emitting code:
I found an unresolved generic type `t2` for the expression `_` at line 220, column 22 in 'D:/Tools/p/carp/core/Macros.carp'
```

So: how does this look for a solution?
```Clojure
  ; keep (System.exit) as it is
  (doc      exit_ "same as exit, but avoids some obscure errors on Windows (e.g. undefined reference to `WinMain'")
  (register exit_ (Fn [Int] ()) "exit")  
```

Reply:
Ok, that's interesting. For my curiosity do you still get linker errors if you do `(ignore (the Int (System.exit 0)))` instead?

I'm reticent having two slightly the same `exit`, there should only be one. Maybe keeping the `System` changes out of this PR would make your life easier.

Reply:
> Ok, that's interesting. For my curiosity do you still get linker errors if you do `(ignore (the Int (System.exit 0)))` instead?
Now _that_ works like a charm.

> I'm reticent having two slightly the same `exit`, there should only be one. Maybe keeping the `System` changes out of this PR would make your life easier.
Let me take a look at the SDL examples to see if we can't just have a non-returning exit function, e.g. (Fn [Int] () ).

Reply:
So it seems we have a bug where if you pass and a function with a generic return at the end of main, we don't generate the main function at all. I would keep the System changes out of this PR and see if we can solve the problem separately. @eriksvedang What do you think?

Reply:
Ah, I see. Yeah that would make `main` generic which makes it not be emitted. Seems like a separate issue yes (where we should detect that `main` has become generic and warn about that.)

Reply:
> Ah, I see. Yeah that would make `main` generic which makes it not be emitted. Seems like a separate issue yes (where we should detect that `main` has become generic and warn about that.)

Oh well, I opened request #1246 to handle the problem separately.

However I would prefer wrapping the code in Main in an implicit ```(ignore (the Int ....))``` over displaying a warning.

Reply:
Warning makes more sense because the two thing we want to accept for `main` is either `()` which means an exit code of 0, or explicitly return an exit code as an `Int`.

It seems the reason you're getting the error is because you're trying to do something like `(System.exit 42)` when you can just replace that with `42` to get the same behaviour.

Reply:
> Warning/Error makes more sense because the two thing we want to accept for `main` is either `()` which means an exit code of 0, or explicitly return an exit code as an `Int`.

Ah, thanks for the pointer
I added it to the list at #1029.